src/components/__test__/helper.spec.ts

```diff
@@ -0,0 +1,170 @@
+  deleteRule,
+  filterObject,
+  setDefaultValue,
+  updateRulesList,
+  validateInput,
+} from '../../helper';
+import { QueryObject } from '../../types';
+
+describe('ValidateInput', () => {
+  it('returns false for wrong input value and the correct errorMessage', () => {
+    const queryObject: QueryObject = {
+      id: '1',
+      combinator: 'AND',
+      rules: [
+        {
+          id: '1',
+          field: 'Age',
+          operator: '=',
+          value: '12d',
+          placeHolder: 'E.g 10',
+          isValid: true,
+          errorMessage: '',
+          residentId: '2',
+          nonResidentId: '3',
+        },
+        {
+          id: '2',
+          field: 'Last Name',
+          operator: '=',
+          value: 'Doe2',
+          placeHolder: 'E.g Doe',
+          isValid: true,
+          errorMessage: '',
+          residentId: '4',
+          nonResidentId: '5',
+        },
+        {
+          id: '3',
+          field: 'Enrollment Year',
+          operator: '=',
+          value: '1993e',
+          placeHolder: 'E.g 2023',
+          isValid: true,
+          errorMessage: '',
+          residentId: '6',
+          nonResidentId: '7',
+        },
+      ],
+    };
+
+    const ageTestObject = validateInput(queryObject, '12d', '1');
+    const nameTestObject = validateInput(queryObject, 'Doe2', '2');
+    const enrollmentYearTestObject = validateInput(queryObject, '1993e', '3');
+
+    expect(ageTestObject).toEqual({
+      isValid: false,
+      errorMessage: 'Please enter a valid age using digits only',
+    });
+    expect(nameTestObject).toEqual({
+      isValid: false,
+      errorMessage: 'Name should contain alphabetical characters only',
+    });
+    expect(enrollmentYearTestObject).toEqual({
+      isValid: false,
+      errorMessage: 'Please enter a valid year using digits only',
+    });
+  });
+});
+
+describe('DeleteRule', () => {
+  it('deletes a rule', () => {
+    const queryObject: QueryObject = {
+      id: '1',
+      combinator: 'AND',
+      rules: [
+        {
+          id: '1',
+          field: 'Level',
+          operator: '=',
+          value: 'Grade I',
+          placeHolder: 'E.g Grade II',
+          isValid: true,
+          errorMessage: '',
+          residentId: '4',
+          nonResidentId: '5',
+        },
+      ],
+    };
+    deleteRule(queryObject, '1');
+    expect(queryObject.rules.length).toEqual(0);
+  });
+});
+
+describe('UpdateRulesList', () => {
+  it('adds a new rule to the rules list', () => {
+    const queryObject: QueryObject = {
+      id: '1',
+      combinator: 'AND',
+      rules: [],
+    };
+    updateRulesList(queryObject);
+    expect(queryObject.rules.length).toEqual(1);
+  });
+});
+
+describe('ChangeInputPlaceHolder', () => {
+  it('returns the right placeholder on field toggle', () => {
+    const placeHolder = changeInputPlaceHolder('Last Name');
+    expect(placeHolder).toEqual('E.g Doe');
+  });
+});
+
+describe('FilterObject', () => {
```

Comment:
Sorry this has come late, but I would suggest changing the filterObject function. the way it is currently written, is not easily testable, it has a lot of de-structuring which makes it somewhat complex. Re-write it to take in the queryObject and  a list of requiredFields as arguments. This way, testing is clearer and also the function can be used elsewhere incase you ever need other fields returned instead.